testing/integration/otel_test.go

```diff
@@ -0,0 +1,295 @@
+
+	require.Eventually(t,
+		func() bool {
+			// verify file exists
+			content, err := os.ReadFile(fileProcessingFilename)
+			return err == nil && len(content) > 0
+		},
+		5*time.Minute, 500*time.Millisecond,
+		"there should be exported logs by now")
+
+	cancel()
+	fixtureWg.Wait()
+	require.True(t, err == nil || err == context.Canceled || err == context.DeadlineExceeded, "Retrieved unexpected error: %s", err.Error())
+}
+
+func TestAPMIngestion(t *testing.T) {
+	info := define.Require(t, define.Requirements{
+		Group: Default,
+		Stack: &define.Stack{},
+		Local: true,
+		OS: []define.OS{
+			// apm server not supported on darwin
+			{Type: define.Linux},
+		},
+	})
+
+	// prepare agent
+	fixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+
+	ctx, cancel := testcontext.WithDeadline(t, context.Background(), time.Now().Add(10*time.Minute))
+	defer cancel()
+	err = fixture.Prepare(ctx, fakeComponent, fakeShipper)
+	require.NoError(t, err)
+
+	// prepare input
+	agentWorkDir := fixture.WorkDir()
+	fileName := "content.log"
+	fixture.WriteFileToWorkDir(ctx, "", fileName)
+
+	testUuid, err := uuid.NewV4()
+	require.NoError(t, err, "failed to create test id")
+	testId := testUuid.String()
+
+	apmConfig := fmt.Sprintf(apmOtelConfig, filepath.Join(agentWorkDir, fileName), testId)
+
+	err = fixture.Configure(ctx, []byte(apmConfig))
+	require.NoError(t, err)
+
+	componentsDir, err := aTesting.FindComponentsDir(agentWorkDir)
+	require.NoError(t, err)
+
+	// start apm default config just configure ES output
+	esHost, err := getESHost()
+	require.NoError(t, err, "failed to get ES host")
+	require.True(t, len(esHost) > 0)
+
+	esUsername := os.Getenv("ELASTICSEARCH_USERNAME")
+	require.True(t, len(esUsername) > 0)
+
+	esPass := os.Getenv("ELASTICSEARCH_PASSWORD")
+	require.True(t, len(esPass) > 0)
+
+	esClient := info.ESClient
+	if esClient == nil {
+		esClient, err = getESClient(esHost, esUsername, esPass)
+		require.NoError(t, err, "failed to create Elasticsearch client")
+	}
+
+	esApiKey, err := getESApiKey(esClient, esHost, esUsername, esPass)
+	require.NoError(t, err, "failed to get api key")
+	require.True(t, len(esApiKey) > 1, "api key is invalid %q", esApiKey)
+
+	apmArgs := []string{
+		"run",
+		"-e",
+		"-E", "output.elasticsearch.hosts=['" + esHost + "']",
+		"-E", "output.elasticsearch.api_key=" + esApiKey,
+		"-E", "apm-server.host=127.0.0.1:8200",
+		"-E", "apm-server.ssl.enabled=false",
+	}
+
+	apmPath := filepath.Join(componentsDir, "apm-server")
+	var apmFixtureWg sync.WaitGroup
+	apmFixtureWg.Add(1)
+	apmContext, apmCancel := context.WithCancel(ctx)
+	defer apmCancel()
+	go func() {
+		aTesting.RunProcess(t,
+			apmContext, 0,
+			true, true,
+			apmPath, apmArgs...)
+		apmFixtureWg.Done()
+	}()
+
+	// start agent
+	var fixtureWg sync.WaitGroup
+	fixtureWg.Add(1)
+	go func() {
+		fixture.RunWithClient(ctx, false)
+		fixtureWg.Done()
+	}()
+
+	go func() {
+		// delayed write
+		<-time.After(30 * time.Second)
```

Comment:
Why do we need this delay? Does it have something to do with Agent starting up in collector mode first or is there some other reason?

Thirty seconds is a sizeable delay IMO. If we could replace this with blocking on a receive from a channel or something like a wait-for pattern that polls for something every second up to 30 seconds, that would be better.

Reply:
yes this is just an arbitrary number of seconds we let otel to work, detect file .... we don't have a way of being informed about processing being done atm.